### PR TITLE
1335 add get regulated bus api

### LINKED ISSUE #1335: Add getRegulatedBus method in data interface services
method to get the regulated bus of a regulating object (e.g. generator, shunt, svc, ..)

### PR DESCRIPTION
**Pull request by marcochiaramello**
Closes #1335

### CODE REVIEW

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 1]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+    case ComponentInterface::THREE_WTFO:
+    case ComponentInterface::VSC_CONVERTER:
+    case ComponentInterface::LCC_CONVERTER:
+    case ComponentInterface::HVDC_LINE:
+      break;
+  }
+  return boost::shared_ptr<BusInterface> ();
+}
+
+boost::shared_ptr<BusInterface>
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:134]: const &

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 2]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+ServiceManagerInterfaceIIDM::getRegulatedBusOnSide(const powsybl::iidm::Terminal& terminal) const {
+  const auto& regulatedComponent = dataInterface_->findComponent(terminal.getConnectable().get().getId());
+  switch (regulatedComponent->getType()) {
+    case ComponentInterface::UNKNOWN:
+    case ComponentInterface::HVDC_LINE:
+      break;
+    case ComponentInterface::LINE: {
+      boost::shared_ptr<LineInterface> line = boost::dynamic_pointer_cast<LineInterface>(regulatedComponent);
+      auto& lineIIDM = dataInterface_->getNetworkIIDM().getLine(regulatedComponent.get()->getID());
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:143]: const line

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 3]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+        return line.get()->getBusInterface1();
+      return line.get()->getBusInterface2();
+    }
+    case ComponentInterface::BUS:
+    case ComponentInterface::CALCULATED_BUS: {
+      return boost::dynamic_pointer_cast<BusInterface>(regulatedComponent);
+    }
+    case ComponentInterface::SWITCH: {
+      boost::shared_ptr<SwitchInterface> switch_ = boost::dynamic_pointer_cast<SwitchInterface>(regulatedComponent);
+      const auto SwitchTerminals = terminal.getConnectable().get().getTerminals();
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:154]: missing reference

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 4]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+    }
+    case ComponentInterface::SHUNT: {
+      return boost::dynamic_pointer_cast<ShuntCompensatorInterface>(regulatedComponent).get()->getBusInterface();
+    }
+    case ComponentInterface::DANGLING_LINE: {
+      return boost::dynamic_pointer_cast<DanglingLineInterface>(regulatedComponent).get()->getBusInterface();
+    }
+    case ComponentInterface::TWO_WTFO: {
+      boost::shared_ptr<TwoWTransformerInterface> TwoWTransf = boost::dynamic_pointer_cast<TwoWTransformerInterface>(regulatedComponent);
+      auto& TwoWTransfIIDM = dataInterface_->getNetworkIIDM().getTwoWindingsTransformer(regulatedComponent.get()->getID());
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:171]: const

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 5]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+    }
+    case ComponentInterface::LCC_CONVERTER: {
+      return boost::dynamic_pointer_cast<LccConverterInterface>(regulatedComponent).get()->getBusInterface();
+    }
+    case ComponentInterface::GENERATOR: {
+      return boost::dynamic_pointer_cast<GeneratorInterface>(regulatedComponent).get()->getBusInterface();
+    }
+    case ComponentInterface::THREE_WTFO: {
+      boost::shared_ptr<ThreeWTransformerInterface> ThreeWTransf = boost::dynamic_pointer_cast<ThreeWTransformerInterface>(regulatedComponent);
+      auto& ThreeWTransfIIDM = dataInterface_->getNetworkIIDM().getThreeWindingsTransformer(regulatedComponent.get()->getID());
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:190]: const

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp [hunk 6]
```diff
@@ -139,4 +171,311 @@ TEST(DataInterfaceTest, ServiceManager) {
...
+  ASSERT_EQ(serviceManager->getRegulatedBus(genItf3.getID())->getID(), bus1->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(genItf4.getID())->getID(), bus5->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(genItf5.getID())->getID(), bus5->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(genItf6.getID())->getID(), bus1->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(genItf7.getID())->getID(), bus1->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(shuntItf2.getID())->getID(), bus5->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(shuntItf.getID())->getID(), bus3->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(svcItf.getID())->getID(), bus1->getID());
+  ASSERT_EQ(serviceManager->getRegulatedBus(svcItf2.getID())->getID(), bus1->getID());
+  serviceManager->getRegulatedBus(lineItf.getID());
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp:478]: missing test?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h [hunk 7]
```diff
@@ -47,6 +47,14 @@ struct ServiceManagerInterface {
    * @returns the list of the bus ids linked to @p busId
    */
   virtual std::vector<std::string> getBusesConnectedBySwitch(const std::string& busId, const std::string& VLId) const = 0;
+
+  /**
+   * @brief Retrive the regulated bus interface
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h:52]: thanks

File: dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h [hunk 8]
```diff
@@ -47,6 +47,14 @@ struct ServiceManagerInterface {
...
+   *
+   * @param regulatingComponent id of the regulating object
+   * @returns regualted bus interface
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h:55]: fixed

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp (hunk 6 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp:478]: yep, fixed, thanks

File: dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h [hunk 9]
```diff
@@ -47,6 +47,14 @@ struct ServiceManagerInterface {
...
+   */
+  virtual boost::shared_ptr<BusInterface> getRegulatedBus(const std::string& regulatingComponent) const = 0;
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h:57]: Not for pointer for me

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.h [hunk 10]
```diff
@@ -24,6 +24,7 @@
 
 #include <string>
 #include <vector>
+#include <boost/unordered_set.hpp>
```
Comment (lecourtoisflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.h:27]: useless header

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.h [hunk 11]
```diff
@@ -177,6 +183,14 @@ class DataInterfaceIIDM : public DataInterface {
    */
   boost::shared_ptr<ComponentInterface>& findComponent(const std::string& id);
 
+  /**
+   * @brief find a component thanks to its id
+   * @param id : id of the component to find
+   *
+   * @return instance of component interface found
+   */
+  const boost::shared_ptr<ComponentInterface>& findComponent(const std::string& id) const;
```
Comment (lecourtoisflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.h:192]: For pointer it is better to use a copy for me (same performance cost as we copy a c pointer

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 12]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+    case ComponentInterface::TWO_WTFO:
+    case ComponentInterface::CALCULATED_BUS:
+    case ComponentInterface::SWITCH:
+    case ComponentInterface::LOAD:
+    case ComponentInterface::LINE:
+    case ComponentInterface::DANGLING_LINE:
+    case ComponentInterface::THREE_WTFO:
+    case ComponentInterface::VSC_CONVERTER:
+    case ComponentInterface::LCC_CONVERTER:
+    case ComponentInterface::HVDC_LINE:
```
Comment (lecourtoisflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:128]: maybe add a comment to explain why only these 3 cases generate a regulated interface

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 13]
```diff
@@ -93,4 +98,104 @@ ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId,
...
+      break;
```
Comment (lecourtoisflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:129]: Maybe replace all these cases by "default" case

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 1 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:134]: not for pointer for me

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.h [hunk 14]
```diff
@@ -58,6 +63,14 @@ class ServiceManagerInterfaceIIDM : public ServiceManagerInterface {
    */
   static void buildGraph(Graph& graph, const boost::shared_ptr<VoltageLevelInterface>& vl);
 
+  /**
+   * @brief retrieve the regulated bus on the side of the regulated object
+   *
+   * @param terminal regualted terminal
```
Comment (lecourtoisflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.h:69]: typo

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h:57]: it is equivalent so it is a matter of personal preference I guess. Anyway in this case we have to copy it as we return a null ptr when the regulated bus ws not found.

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 15]
```diff
@@ -19,6 +19,7 @@
  */
 
 #include "DYNServiceManagerInterfaceIIDM.h"
+#include "DYNBusInterface.h"
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.cpp:22]: indeed

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 2 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:143]: done

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 3 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:154]: done

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 13 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:129]: forced to be explicit by compilation

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.h (hunk 11 above)
Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.h:192]: I'd keep this as it is to have symmetry with respect to other declaration of the method

**Review by lecourtoisflo — APPROVED** [review]

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.cpp:22]: still there!

**Review by rosiereflo — CHANGES_REQUESTED** [review]

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**